_posts/2022-10-05-pdf-generation-at-scale.md

```diff
@@ -0,0 +1,361 @@
+      console.log(`Saved as: ${key}`)
+
+      return context.status(200).succeed()
+    }
+
+    return context.status(400).fail()
+  }
+
+  let body = JSON.parse(event.body)
+  let pages = body.pages
+  console.log("Submitting pages")
+
+  pages.forEach(async (uri) => {
+    const res = await axios.post(
+      "http://gateway.openfaas:8080/async-function/page-to-pdf",
+      { uri },
+      { headers: { 'X-Callback-Url': 'http://gateway.openfaas:8080/function/pdf-gen/save' } }
+    )
+  })
+
+  return context
+    .status(202)
+    .succeed()
+}
+```
+
+The function invokes `page-to-pdf` asynchronously and sets the `X-Callback-Url` header to receive the result once it has finished processing. The reason we want to invoke the functions using its async endpoint is so that we can make use of the built in retry mechanism of OpenFaaS. We will see how this works in the next sections.
+
+To try out the workflow save some web page urls in json file, `pages.json`, and invoke the `gen-pdf` function with `curl`:
+
+```json
+{
+  "pages": [
+    "https://docs.openfaas.com"
+    "https://docs.inlets.dev",
+  ]
+}
+```
+
+```bash
+curl -i $OPENFAAS_URL/function/pdf-gen -d @pages.json -H "Content-Type: application/json"
+
+HTTP/1.1 202 Accepted
+Content-Length: 0
+Content-Type: application/json
+Date: Wed, 05 Oct 2022 11:57:42 GMT
+X-Call-Id: 39d0d02c-491d-4833-b7ea-633e289e01a6
+X-Duration-Seconds: 0.032700
+X-Start-Time: 1664971062806019972
+```
+
+When you look at the logs for the `gen-pdf` function you can see the pages are submitted and the save endpoint is called with the generated PDF for each url.
+
+```
+pdf-gen-d7c57ff8b-lp8s8 pdf-gen Submitting pages
+pdf-gen-d7c57ff8b-lp8s8 pdf-gen 2022/10/05 11:57:42 POST / - 202 Accepted - ContentLength: 0
+pdf-gen-d7c57ff8b-lp8s8 pdf-gen Received pdf of: https://docs.openfaas.com
+pdf-gen-d7c57ff8b-lp8s8 pdf-gen Saved as: 57e02fedbbdc98a8449193ee4a5268c7ccf6bf113bc64ce4d34118775e8412b9.pdf
+pdf-gen-d7c57ff8b-lp8s8 pdf-gen 2022/10/05 11:57:45 POST /save - 200 OK - ContentLength: 0
+pdf-gen-d7c57ff8b-lp8s8 pdf-gen Received pdf of: https://docs.inlets.dev
+pdf-gen-d7c57ff8b-lp8s8 pdf-gen Saved as: b9f84c4e28ac597bef927b1ec3411c4e22056ebd3a06b1ff20a81cf919edf17f.pdf
+pdf-gen-d7c57ff8b-lp8s8 pdf-gen 2022/10/05 11:57:48 POST /save - 200 OK - ContentLength: 0
+```
+
+You should be able to find the PDFs using the AWS S3 console.
+
+![AWS Console with generated PDF](/images/2022-pdf-generation-at-scale/aws-console-and-pdf.png)
+> AWS S3 Console with generated PDF
+
+## Prevent overloading the function
+
+Each replica of the PDF generation function can only run N browsers at once. To prevent overloading it we can set a hard limit on the number of concurrent requests the function can handle. We can for example set a limit of 1 by setting the `max_inflight` environment variable. Any subsequent requests would be dropped and receive a 429 response. This assumes the producer can buffer the requests to retry them later on.
```

Comment:
Consider swapping `can only` to `will` here

Reply:
My intention is to make it clear that the function is only capable of running N browsers at once. Not sure replacing `can only` with `will` makes that clear.